src/TypeAdapter/ReflectionTypeAdapter.php

```diff
@@ -262,77 +248,95 @@ public function read(JsonReadable $reader)
     /**
      * Write the value to the writer for the type
      *
-     * @param JsonWritable $writer
      * @param object $value
-     * @return void
+     * @param WriterContext $context
+     * @return array|null
      */
-    public function write(JsonWritable $writer, $value): void
+    public function write($value, WriterContext $context): ?array
     {
         if ($this->skipSerialize || $value === null) {
-            $writer->writeNull();
-            return;
+            return null;
         }
 
         $classExclusionCheck = $this->hasClassSerializationStrategies
-            && (!$this->requireExclusionCheck || ($this->requireExclusionCheck && $this->classMetadata->getAnnotation(ExclusionCheck::class) !== null));
+            && (!$this->requireExclusionCheck || ($this->requireExclusionCheck && $this->classAnnotations->get(ExclusionCheck::class) !== null));
         $propertyExclusionCheck = $this->hasPropertySerializationStrategies
             && (!$this->requireExclusionCheck || ($this->requireExclusionCheck && $this->hasPropertyExclusionCheck));
         $exclusionData = $classExclusionCheck || $propertyExclusionCheck
-            ? new DefaultSerializationExclusionData($value, $writer)
+            ? new DefaultSerializationExclusionData($value, $context)
             : null;
 
         if ($classExclusionCheck && $exclusionData) {
             $this->excluder->applyClassSerializationExclusionData($exclusionData);
 
             if ($this->excluder->excludeClassBySerializationStrategy($this->classMetadata)) {
-                $writer->writeNull();
-                return;
+                return null;
             }
         }
 
         if ($propertyExclusionCheck && $exclusionData) {
             $this->excluder->applyPropertySerializationExclusionData($exclusionData);
         }
 
-        $writer->beginObject();
-
-        if ($this->classVirtualProperty !== null) {
-            $writer->name($this->classVirtualProperty);
-            $writer->beginObject();
-        }
+        $enableScalarAdapters = $context->enableScalarAdapters();
+        $serializeNull = $context->serializeNull();
+        $result = [];
 
         /** @var Property $property */
         foreach ($this->properties as $property) {
-            $writer->name($property->getSerializedName());
-            if ($property->skipSerialize()) {
-                $writer->writeNull();
+            $serializedName = $property->serializedName;
+
+            if ($property->skipSerialize) {
```

Comment:
Could this be evaluated before `$serializedName`?

Reply:
yeah